scrapy/core/engine.py

```diff
@@ -120,24 +149,23 @@ def _next_request(self, spider):
         if self.paused:
             return
 
-        while not self._needs_backout(spider):
-            if not self._next_request_from_scheduler(spider):
-                break
-
-        if slot.start_requests and not self._needs_backout(spider):
-            try:
-                request = next(slot.start_requests)
-            except StopIteration:
-                slot.start_requests = None
-            except Exception:
-                slot.start_requests = None
-                logger.error('Error while obtaining start requests',
-                             exc_info=True, extra={'spider': spider})
+        if self._waiting_for_request:
+            return
+        self._waiting_for_request = True
+        try:
+            if slot.new_queue_behavior:
+                while not self._needs_backout(spider) and slot.start_requests:
+                    yield self._schedule_next_req(spider, slot)
+                self._send_reqs_to_downloader(spider)
```

Comment:
What happens in examples like that?

```py
async def start_requests(self):
    # yield many requests
    await asyncio.sleep(10)
    # yield many requests
```

I'd expect that

1) while start_request is sleeping, requests scheduled before that are being downloaded;
2) if responses are received during this 10s sleep, callbacks are executed as usual, which may result in items extracted, new requests scheduled, etc.;
3) when start_request resumes after `await`, remaining requests are put to the scheduler as usual, without waiting  for any queues to become empty

Is all of this working? Sorry if that's a stupid question, I haven't really understood the implementation yet.

Reply:
I've just tested this, and it works in a different way: the requests are scheduled, but `_send_reqs_to_downloader` is only called after `sleep` returns, so nothing is downloaded before that and these use cases cannot be tested.

Reply:
Hm, is it possibile to fix that? I think it breaks some of the most important use cases, like this one:

```py
async def start_requests(self):
    while True:
        batch = await get_next_batch_from_redis()
        for req in batch:
            yield scrapy.Request(req['url'], ...) 
```

> and these use cases cannot be tested.

what do you mean by that?

Reply:
> Hm, is it possibile to fix that?

Without limiting the number of requests scheduled in the while loop? No idea. I can only guess whether the initial idea was to load all available requests into the scheduler before starting to move them to the downloader.  

The only code I added here, `yield deferred_from_coro(slot.start_requests.__anext__())`, doesn't distinguish between a long wait and just getting the next value from the generator.

> what do you mean by that?

I mean this test case cannot give us answers whether your points 2 and 3 are correct (if there are use cases where they make sense).

Reply:
I'm actually unsure what's the use case for `async def start_requests` if the requests don't start to download while something is being awaited in `async def start_requests`. Why not make e.g. a sync time.sleep, or a sync call to redis instead, won't the behavior be the same in the end? 

Do you see a use case for that? If not, I think it would be better to remove 2.4 milestone from this PR, and wait until that is fixed.

There is a use case though where we might need to ensure that all start_requests made it into the scheduler, before starting to download them. I wonder if we can have an API for that as well - maybe even a different method. It may be handled in the current implementation (aiter may be consuming requests and scheduling them without yielding to the event loop, which makes behavior the same), but I'm not sure about that.

A good point about `yield deferred_from_coro(slot.start_requests.__anext__())`. But maybe that's fine to not distinguish between those, not guarantee that downloads don't start until await.

How hard would it be to modify start_requests to ensure downloading happens in parallel?

Reply:
I agree that the current behavior is unacceptable but I don't know what should be the correct behavior. The current code works as it was written in #3237, loading all requests from `start_requests` into the scheduler. 

I'm fine with removing this PR from 2.4.

Reply:
About the `sleep` issue, I wonder it we could iterate over requests from an asynchronous `start_requests` [using a timeout](https://stackoverflow.com/q/50241696/939364), so that if no requests are yielded for a certain time, other stuff is handled for some time, and iteration of `start_requests` is resumed at some later point.

I’m not sure how viable this would be, though, as I’m talking from a complete lack of familiarity with the actual code.